README
==========================
         Keryx
==========================
      Version 0.92

==========================
     Contact/Support:
==========================

	Website: http://keryxproject.org
	Email: Chris Oliver <excid3@gmail.com>
    Launchpad Project: http://launchpad.net/keryx

==========================
         License:
==========================

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

==========================
       Requirements
==========================

Windows: None

Linux: Python, wxPython

Ubuntu users need the following packages in order to have wxPython successfully
installed:

libwxbase2.8-0
libwxgtk2.8-0
python-wxversion
python-wxgtk2.8

wxPython is only required for using Keryx's interface. Project creation can be
done via command line parameters. See USAGE for more information.

==========================
Installation Instructions:
==========================

Extract the Keryx archive anywhere (typically a usb device).

==========================
          Usage:
==========================
Keryx is based around the concept of 'projects'. Each project keeps track of the
software versions that are installed on that computer. To get started you must
first create a project on that computer.

If you are unable to install wxPython but have Python installed you can still
create projects. In a console run:

'python keryx.py --create <project name> <plugin name>'

and Keryx will create a new project in the project directory. This is helpful
for Ubuntu server users along with users who have not installed wxPython yet.
The only <plugin name> currently available is debian.

Once you have created this project, make sure you have a copy on your usb device
that will be used to run Keryx from on a computer with highspeed internet.

Keryx can be run on ANY computer with Python and wxPython installed, meaning
Windows, Mac and Linux computers. Running Keryx on any Windows computer is
nothing more than running 'keryx.exe'. If you are running Keryx using Python, cd
to the directory and run 'python keryx.py' to start. Some environments support
running a python application by double clicking it as well.

After you have Keryx running on an internet connected host, downloading packages
is as simple as opening your project and selecting the packages to download.
Keryx will automatically select the package dependencies for you making sure
that nothing is left behind. Any packages you download are saved to your
project's folder in the 'packages' directory.

When you are finished you can return to your project computer and install the
packages through the usual way.

==========================
Uninstallation/Upgrading:
==========================

Extract the newest version over top, replacing all files.

==========================
         Thanks:
==========================

Buran Ayuthia
John Gleissner
mac9416
jacseen
Kevin Buente
Douglass Clem
Cory Thompson
Mark James - Silk Icon Set - http://www.famfamfam.com/lab/icons/silk/
Tiziano Tissino - ProportionalSplitter
Ingelrest François - Plugin framework base
heylove - Logo is based off areao.42-Icons
Anyone else I forgot to mention

==========================
     Additional Notes:
==========================

Command line parameters:

-h or --help                            Displays this message
-v or --version                         Displays Keryx version number
--create <project name> <plugin name>   Creates a new project in the default
                                          directory with name of <project name>
                                          and type project <plugin name>
--config <file>                         Uses <file> to as configuration file

Due to py2exe's nature, information printed to the console is NOT displayed.
While the command line parameters will still work properly, no indication of
success will be shown in the console. The information will still be written in
the log files for further review.


doc/ChangeLog
Keryx 0.92.4 (15 March 2010)
-------------
Improvements:
  * Added examples to keryx.conf
  * Made proxy options GUI more user-friendly
  * Added Update Status button to GUI
  * Added Install Packages tool

Bugfixes:
  * Fixed bug that prevented project creation on some machines (#510637)
  * Now warns when a compatible version of wxPython is not available (#503056)
  * No longer requires a primary sources.list to open a project (#522436)

Known Bugs:
  * On some Linux systems, parts of the interface do not display correctly (#522529)

Keryx 0.92 (20 February 2009)
-------------
Improvements:
  * Added configuration file support
  * Added Sources editor
  * Added folder options
  * Added quick open to start dialog
  * Added proxy support
  * Added a color map index
  * Added translation support
  * Added options
  * Added automatic update checking
 
Bugfixes:
  * Fixed improper debian package version comparing
  * Checks Windows disk free space
  
Known Bugs:
  * Some dependencies are not packages, but other packages Provide the dependency

Keryx 0.91 (31 December 2008)
-------------
Improvements:
  * wxWidgets interface
  * Speed improvements
  * Threading for a more responsive interface
  * Searching
  * Windows binary    
  * New logo
  
Known Bugs:
  * Windows binary cannot display text printed to console  

Keryx 0.12 (26 August 2008)
-------------
Added:
  * Download.py:
    - Allows overwriting of files if they already exist
    - If exists and not overwriting, notifies user that file already exists
  * Ubuntu.py:
    - Overwrites package lists if 'internet' is selected

Bugfixes:
  * Fixed #259924 in Ubuntu.py
  * Internet package lists are downloaded correctly

Keryx 0.11.1 (19 August 2008)
-------------
Bugfixes:
  * Fixed bug #258912 in Ubuntu.py
  * Packages duplicates are removed
  * Fixed an issue with repositories with only one section

Keryx 0.11 (10 August 2008)
-------------
Added:
  * Ubuntu: 
    - Calculates package dependencies and downloads them
    - Caches package lists to project folder
    - Looks in pkglists folder for Local Files
    - Checks to see if package is already downloaded
    - Installed packages are displayed if package list is empty on project refresh
Bugfixes:
  * Small bugfixes
  * Ubuntu: 
    - Status file moved to pkglists folder
    - Correctly parses deb entries with multiple sections
  
Keryx 0.10 (31 July 2008)
-------------------------
  * Initial Public Release


keryx.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Author: Chris Oliver (excid3@gmail.com)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA

# Import commands here, even though we don't use it because debian.py uses it
# and our plugin system confuses pyinstaller. We need to manually specify this
# in keryx so that pyinstaller can grab it and bundle the code

# Import modules to satisfy plugin dependencies:
import commands

# Then import everything else.
import os
import sys

from lib import consts

#####################
# Set configuration #
#####################
if ('--config' in sys.argv): # If config is specified
    index = sys.argv.index('--config')
    try: filename = sys.argv[index + 1]
    except: 
        print 'ERROR: No config filename given'
        sys.exit(1)
        
    
    # Reinitialize consts if the a specific file has been passed
    import lib.config
    lib.config.fromFile(filename)

# Else if file_config exists, load it
elif os.path.exists(consts.file_config):
    import lib.config
    lib.config.fromFile(consts.file_config)
    
# Import log and project AFTER configuration has been set
from lib import log, project

################
# Translations #
################
try:
    import gettext
    import locale
    gettext.install(consts.appNameShort, consts.dirLocale, unicode=True)
    
    lang = locale.getdefaultlocale()[0]
    lang = gettext.translation(consts.appNameShort, consts.dirLocale, languages=[lang])
    lang.install()
except: pass

#TODO: Set wxWidgets translation information
 
###################
# Parse arguments #
###################
if ('--help' in sys.argv) or ('-h' in sys.argv):
    log.info(consts.parameters)
    sys.exit(0)

if ('--version' in sys.argv) or ('-v' in sys.argv):
    log.info('%s %s' % (consts.appName, consts.appVersion))
    sys.exit(0)

if ('--create' in sys.argv):
    import platform
    from lib import plugins, project
    try:
        index = sys.argv.index('--create')
        name = sys.argv[index + 1]
        plugin_name = 'debian'
    except:
        log.error(_('Unable to create project'))
        sys.exit(1)
    plugins.load(consts.dirPlugins, '', False) # Don't load interface plugins
    for item in range(0,len(plugins.OSPluginList)):
        if plugin_name == plugins.OSPluginList[item][0].lower():
            # Append new project
            project.projects.append(project.Project())
            proj = project.projects[len(project.projects) - 1]
    
            # Create project
            success, filename = proj.CreateKeryx(name, 
                                            plugins.OSPluginList[item][0],
                                            plugins.OSPluginList[item][1])

            if success: 
                log.info(_('Project created successfully.'))
                sys.exit(0)
    log.error(_('Unable to create project.'))
    log.info(_('Make sure a project by this name does not already exist and you have selected the right plugin for this project.'))
    sys.exit(1)


WXVER = "2.8"

if not hasattr(sys, "frozen"):  # If this isn't a compiled version of Keryx...
    import wxversion
    if wxversion.checkInstalled(WXVER):
        wxversion.select(WXVER)
    else:
        import wx
        app = wx.PySimpleApp()
        wx.MessageBox("Warning: The requested version of wxPython is not installed.\n"
                      "Please install version %s" % WXVER,
                      "wxPython Version Warning")
        app.MainLoop()

#############################
# Attempt to load interface #
#############################
#try:
if 1:
    import lib.wxkeryx
    lib.wxkeryx.Start()
#except Exception, e:
    #print e
#    log.error(_('Unable to load interface.'))
#    log.info(_('Make sure you have wxPython installed.'))
#    log.info(_('Read the README for details on installing wxPython'))
#    log.info(_('You can you \'python keryx.py --create\' to create a new project through the command-line if you do not have wxPython installed'))
#    log.info(_('Use \'python keryx.py --help\' to see all command-line parameters.'))
#    sys.exit(1)
